src/App/Fossa/VSI/DynLinked/Internal/Resolve.hs

```diff
@@ -0,0 +1,86 @@
+{-# LANGUAGE RecordWildCards #-}
+
+module App.Fossa.VSI.DynLinked.Internal.Resolve (
+  toSourceUnit,
+  toDependency,
+) where
+
+import App.Fossa.Analyze.Project (ProjectResult (..))
+import App.Fossa.BinaryDeps (analyzeSingleBinary)
+import App.Fossa.VSI.DynLinked.Types (DynamicDependency (..), LinuxPackageManager (..), LinuxPackageMetadata (..), ResolvedLinuxPackage (..))
+import Control.Algebra (Has)
+import Control.Effect.Diagnostics (Diagnostics)
+import Control.Effect.Lift (Lift)
+import Data.Set (Set, toList)
+import Data.Text (intercalate)
+import DepTypes (DepType (LinuxAPK, LinuxDEB, LinuxRPM), Dependency (..), VerConstraint (CEq))
+import Effect.Logger (Logger)
+import Effect.ReadFS (ReadFS)
+import Graphing qualified
+import Path (Abs, Dir, File, Path)
+import Srclib.Converter qualified as Srclib
+import Srclib.Types (AdditionalDepData (..), SourceUnit (..))
+import Types (DiscoveredProjectType (VsiProjectType), GraphBreadth (Complete))
+
+-- | Resolves a set of dynamic dependencies into a @SourceUnit@.
+-- Any @DynamicDependency@ that isn't resolved to a Linux package dependency is converted to an unknown binary dependency.
+toSourceUnit :: (Has (Lift IO) sig m, Has Logger sig m, Has ReadFS sig m, Has Diagnostics sig m) => Path Abs Dir -> Set DynamicDependency -> m SourceUnit
+toSourceUnit root dependencies = do
+  let (resolved, unresolved) = sortResolvedUnresolved dependencies
+  binaries <- traverse (analyzeSingleBinary root) unresolved
+
+  let project = toProject root $ Graphing.fromList (map toDependency resolved)
```

Comment:
[nit] I know `fromList` is alias for directs, but would prefer `directs` to signify they are direct deps.